### ISSUE TITLE
Build error: ‘cudnnSetRNNDescriptor’ was not declared in this scope

### ISSUE THREAD

**Issue by QuentGrd** [facebookresearch/flashlight#147]
### Bug Description
I'm trying to build flashlight from sources with Cuda backend following steps provided in the documentation but an error occurs in "/flashlight/autograd/backend/cuda/CudnnUtils.cpp". The compiler says (see full logs at the end) : 
```
error: ‘cudnnSetRNNDescriptor’ was not declared in this scope; did you mean ‘cudnnSetLRNDescriptor’?
  280 |   CUDNN_CHECK_ERR(cudnnSetRNNDescriptor(
      |                   ^~~~~~~~~~~~~~~~~~~~~
```
#### Reproduction Steps
```
cmake .. -DCMAKE_BUILD_TYPE=Release -DFLASHLIGHT_BACKEND=CUDA

make clean && make -j4
# Or just the concerned file
make flashlight/autograd/backend/cuda/CudnnUtils.o 
```

### Platform and Hardware
OS: Ubuntu 20.04 LTS
kernel: 5.4.0-42-generic
ArrayFire: 3.7.2
CUDA: 11.0
cuDNN: 8.0.1 RC2
cmake: 3.16.3
gcc/g++: 9.3.0 

#### Additional Context
Full logs:

``` 
quentin@machine:~/Documents/flashlight/build$ cmake .. -DCMAKE_BUILD_TYPE=Release -DFLASHLIGHT_BACKEND=CUDA -Wno-dev
-- ArrayFire found (include: /opt/arrayfire/include, library: ArrayFire::afcuda)
-- Could NOT find cereal (missing: cereal_INCLUDE_DIRS) 
-- cereal NOT found. Will download from source
-- CUDA found (library: /usr/local/cuda-11.0/lib64/libcudart_static.a;-pthread;dl;/usr/lib/x86_64-linux-gnu/librt.so include: /usr/local/cuda-11.0/include)
-- CUDNN libname: libcudnn.so;libcudnn.dylib;cudnn64
-- Found CUDNN: (lib: /usr/lib/x86_64-linux-gnu/libcudnn.so include: /usr/include)
-- CUDNN found (library: /usr/lib/x86_64-linux-gnu/libcudnn.so include: /usr/include)
-- Will build flashlight contrib assets.
-- Gloo not found
-- Determining NCCL version from the header file: /usr/include/nccl.h
-- NCCL_MAJOR_VERSION: 2
-- Found NCCL (include: /usr/include, library: /lib/x86_64-linux-gnu/libnccl.so)
-- NCCL found
-- MPI_VERSION found:  3.1
-- MPI_CXX found
-- MPI_CXX compile flags: -pthread
-- MPI_CXX include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
-- MPI_CXX LINK flags path: -pthread
-- MPI_CXX libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
-- MPI_C found
-- MPI_C compile flags: -pthread
-- MPI_C include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
-- MPI_C LINK flags path: -pthread
-- MPI_C libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
-- Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR GTEST_MAIN_LIBRARY) (Required is at least version "1.10.0")
-- googletest not found - will download and build from source
-- Configuring done
-- Generating done
-- Build files have been written to: /home/quentin/Documents/flashlight/build
```

``` 
quentin@machine:~/Documents/flashlight/build$ make clean && make -j4
[  0%] Creating directories for 'cereal'
[  0%] Creating directories for 'gtest'
[  0%] Performing download step (git clone) for 'cereal'
[  0%] Performing download step (git clone) for 'gtest'
-- Avoiding repeated git clone, stamp file is up to date: '/home/quentin/Documents/flashlight/build/cereal/src/cereal-stamp/cereal-gitclone-lastrun.txt'
-- Avoiding repeated git clone, stamp file is up to date: '/home/quentin/Documents/flashlight/build/tests/googletest/src/gtest-stamp/gtest-gitclone-lastrun.txt'
[  0%] No patch step for 'cereal'
[  0%] Performing update step for 'cereal'
[  0%] Performing update step for 'gtest'
[  0%] No patch step for 'gtest'
[  0%] Performing configure step for 'cereal'
loading initial cache file /home/quentin/Documents/flashlight/build/cereal/tmp/cereal-cache-Release.cmake
[  0%] Performing configure step for 'gtest'
-- Configuring done
-- Generating done
-- Build files have been written to: /home/quentin/Documents/flashlight/build/cereal/src/cereal
loading initial cache file /home/quentin/Documents/flashlight/build/tests/googletest/tmp/gtest-cache-Release.cmake
[  0%] Performing build step for 'cereal'
make[3]: avertissement : jobserver n'est pas disponible : utilisation de -j1. Ajouter « + » à la règle parent du make.
[  0%] No install step for 'cereal'
[  0%] Completed 'cereal'
-- Configuring done
-- Generating done
-- Build files have been written to: /home/quentin/Documents/flashlight/build/tests/googletest/src/gtest
[  0%] Built target cereal
[  1%] Performing build step for 'gtest'
make[3]: avertissement : jobserver n'est pas disponible : utilisation de -j1. Ajouter « + » à la règle parent du make.
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/Functions.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/Variable.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/Utils.cpp.o
[ 25%] Built target gtest
[ 50%] Built target gmock
[ 75%] Built target gmock_main
[100%] Built target gtest_main
[  2%] No install step for 'gtest'
[  2%] Completed 'gtest'
[  2%] Built target gtest
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/BatchNorm.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/Conv2D.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/CudnnUtils.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/Pool2D.cpp.o
In file included from /home/quentin/Documents/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp:9:
/home/quentin/Documents/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp: In constructor ‘fl::RNNDescriptor::RNNDescriptor(af::dtype, int, int, fl::RnnMode, bool, fl::DropoutDescriptor&)’:
/home/quentin/Documents/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp:280:19: error: ‘cudnnSetRNNDescriptor’ was not declared in this scope; did you mean ‘cudnnSetLRNDescriptor’?
  280 |   CUDNN_CHECK_ERR(cudnnSetRNNDescriptor(
      |                   ^~~~~~~~~~~~~~~~~~~~~
/home/quentin/Documents/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.h:101:52: note: in definition of macro ‘CUDNN_CHECK_ERR’
  101 | #define CUDNN_CHECK_ERR(expr) ::fl::cudnnCheckErr((expr))
      |                                                    ^~~~
make[2]: *** [CMakeFiles/flashlight.dir/build.make:128 : CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/CudnnUtils.cpp.o] Erreur 1
make[2]: *** Attente des tâches non terminées....
make[1]: *** [CMakeFiles/Makefile2:197 : CMakeFiles/flashlight.dir/all] Erreur 2
make: *** [Makefile:141 : all] Erreur 2

```

**Comment by jacobkahn** [comment]
Fixed.

**Comment by abhinavkulkarni** [comment]
@jacobkahn, @tlikhomanenko : Was this change added to `master` branch only? I am trying to replicate instructions in [`Docerfile-CUDA`](https://github.com/facebookresearch/wav2letter/blob/v0.2/Dockerfile-CUDA#L15-L18) from `wav2letter++` project which installs `v0.2` version of `flashlight` and I am having the following trouble:

```bash
~$ cat /usr/lib/cuda/version.txt 
CUDA Version 10.1.243

~$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243

~$ sudo apt-get install libcudnn8-dev 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libcudnn8-dev is already the newest version (8.0.5.39-1+cuda11.1).
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.

~$ cd ~/Projects && git clone --recursive https://github.com/facebookresearch/flashlight.git && cd ~/Projects/flashlight && git checkout v0.2 && mkdir -p build && cd build
Cloning into 'flashlight'...
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 16023 (delta 0), reused 4 (delta 0), pack-reused 16011
Receiving objects: 100% (16023/16023), 7.52 MiB | 2.76 MiB/s, done.
Resolving deltas: 100% (11397/11397), done.
Branch 'v0.2' set up to track remote branch 'v0.2' from 'origin'.
Switched to a new branch 'v0.2'

~/Projects/flashlight/build$ cmake .. -DCMAKE_BUILD_TYPE=Release -DFLASHLIGHT_BACKEND=CUDA
-- ArrayFire found (include: /usr/local/include, library: ArrayFire::afcuda)
-- Could NOT find cereal (missing: cereal_INCLUDE_DIRS) 
-- cereal NOT found. Will download from source
-- CUDA found (library: /usr/lib/x86_64-linux-gnu/libcudart_static.a;-pthread;dl;/usr/lib/x86_64-linux-gnu/librt.so include: /usr/include)
-- CUDNN libname: libcudnn.so;libcudnn.dylib;cudnn64
-- Found CUDNN: (lib: /usr/lib/x86_64-linux-gnu/libcudnn.so include: /usr/include)
-- CUDNN found (library: /usr/lib/x86_64-linux-gnu/libcudnn.so include: /usr/include)
-- Will build flashlight contrib assets.
CMake Warning (dev) at flashlight/contrib/modules/CMakeLists.txt:18 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "ContribModules" has a relative path.
Call Stack (most recent call first):
  flashlight/contrib/CMakeLists.txt:12 (include)
  CMakeLists.txt:243 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/autograd/CMakeLists.txt:16 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Autograd" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:279 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/autograd/CMakeLists.txt:79 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Autograd" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:279 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/common/CMakeLists.txt:25 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Common" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:282 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/dataset/CMakeLists.txt:25 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Dataset" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:285 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/distributed/CMakeLists.txt:19 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Distributed" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:288 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Gloo not found
-- Determining NCCL version from the header file: /usr/include/nccl.h
-- NCCL_MAJOR_VERSION: 2
-- Found NCCL (include: /usr/include, library: /lib/x86_64-linux-gnu/libnccl.so)
-- NCCL found
-- MPI_VERSION found:  3.1
-- MPI_CXX found
-- MPI_CXX compile flags: -pthread
-- MPI_CXX include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
-- MPI_CXX LINK flags path: -pthread
-- MPI_CXX libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
-- MPI_C found
-- MPI_C compile flags: -pthread
-- MPI_C include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
-- MPI_C LINK flags path: -pthread
-- MPI_C libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
CMake Warning (dev) at flashlight/memory/CMakeLists.txt:17 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Memory" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:291 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/meter/CMakeLists.txt:18 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Meter" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:294 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/nn/CMakeLists.txt:45 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "NN" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:297 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at flashlight/optim/CMakeLists.txt:22 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "Optim" has a relative path.
Call Stack (most recent call first):
  CMakeLists.txt:300 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR GTEST_MAIN_LIBRARY) (Required is at least version "1.10.0")
-- googletest not found - will download and build from source
-- Configuring done
-- Generating done
-- Build files have been written to: /home/w2luser/Projects/flashlight/build

~/Projects/flashlight/build$ make -j$(nproc)
[  0%] Performing update step for 'cereal'
[  0%] Performing update step for 'gtest'
[  0%] Performing configure step for 'gtest'
[  0%] Performing configure step for 'cereal'
loading initial cache file /home/w2luser/Projects/flashlight/build/cereal/tmp/cereal-cache-Release.cmake
-- Configuring done
-- Generating done
-- Build files have been written to: /home/w2luser/Projects/flashlight/build/cereal/src/cereal
[  0%] Performing build step for 'cereal'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
loading initial cache file /home/w2luser/Projects/flashlight/build/tests/googletest/tmp/gtest-cache-Release.cmake
-- Configuring done
-- Generating done
-- Build files have been written to: /home/w2luser/Projects/flashlight/build/tests/googletest/src/gtest
[  1%] Performing build step for 'gtest'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[  1%] No install step for 'cereal'
[ 25%] Built target gtest
[  1%] Completed 'cereal'
[ 50%] Built target gmock
[ 75%] Built target gmock_main
[  1%] Built target cereal
[100%] Built target gtest_main
[  1%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/ShuffleDataset.cpp.o
[  1%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/PrefetchDataset.cpp.o
[  1%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/MergeDataset.cpp.o
[  1%] Building CXX object CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/CudnnUtils.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/ResampleDataset.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/TransformDataset.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/dataset/TensorDataset.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/distributed/reducers/InlineReducer.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/distributed/DistributedApi.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/distributed/FileStore.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/distributed/reducers/CoalescingReducer.cpp.o
[  2%] No install step for 'gtest'
[  2%] Completed 'gtest'
[  2%] Built target gtest
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/distributed/backend/cuda/DistributedBackend.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/memory/MemoryManagerAdapter.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/memory/MemoryManagerInstaller.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/memory/managers/DefaultMemoryManager.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/memory/managers/CachingMemoryManager.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/meter/AverageValueMeter.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/meter/CountMeter.cpp.o
In file included from /home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp:9:0:
/home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp: In constructor ‘fl::RNNDescriptor::RNNDescriptor(af::dtype, int, int, fl::RnnMode, bool, fl::DropoutDescriptor&)’:
/home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp:280:19: error: ‘cudnnSetRNNDescriptor’ was not declared in this scope
   CUDNN_CHECK_ERR(cudnnSetRNNDescriptor(
                   ^
/home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.h:101:52: note: in definition of macro ‘CUDNN_CHECK_ERR’
 #define CUDNN_CHECK_ERR(expr) ::fl::cudnnCheckErr((expr))
                                                    ^~~~
/home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.cpp:280:19: note: suggested alternative: ‘cudnnSetLRNDescriptor’
   CUDNN_CHECK_ERR(cudnnSetRNNDescriptor(
                   ^
/home/w2luser/Projects/flashlight/flashlight/autograd/backend/cuda/CudnnUtils.h:101:52: note: in definition of macro ‘CUDNN_CHECK_ERR’
 #define CUDNN_CHECK_ERR(expr) ::fl::cudnnCheckErr((expr))
                                                    ^~~~
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/meter/EditDistanceMeter.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/meter/FrameErrorMeter.cpp.o
[  2%] Building CXX object CMakeFiles/flashlight.dir/flashlight/meter/MSEMeter.cpp.o
make[2]: *** [CMakeFiles/flashlight.dir/build.make:128: CMakeFiles/flashlight.dir/flashlight/autograd/backend/cuda/CudnnUtils.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:197: CMakeFiles/flashlight.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

Thanks!

**Comment by abhinavkulkarni** [comment]
I managed to solve the issue by installing `libcudnn7` and `libcudnn7-dev` instead of v8 (which is the latest version in Ubuntu repository and on NVDIA downloads website).

However, I would still appreciate it if the FB team can clarify whether the latest change in branch `v0.2` is supposed to work with v8 or not.

Thanks!

**Comment by tlikhomanenko** [comment]
@abhinavkulkarni Sorry, but we are not going to update branch v0.2 as we have a lot of changes/improvements/updates in current master. I have committed several bug fixes in v0.2 only for emergency fixes so that people can continue experiments with v0.2 while we are preparing new stable master. Please wait a new stable version of master where we have upgrades for cmake, cuda, cudnn, a lot of bug fixes, improvement of API, new layers/models, vcpkg support and a lot of other things =) .

**Comment by abhinavkulkarni** [comment]
@tlikhomanenko: Thanks.

By the time stable `master` is available, what is recommended in the interim for `wav2letter++`? Use `v0.2` or current `master`?
